Clearing a text input and typing in it again causes a panic

```
panic: runtime error: slice bounds out of range

goroutine 1 [running, locked to thread]:
fyne.io/fyne/widget.(*textProvider).insertAt(0xc4201cc2a0, 0x1, 0xc420af9a64, 0x1, 0x1)
	/Users/pbarnum/go/src/fyne.io/fyne/widget/text.go:147 +0x26f
fyne.io/fyne/widget.(*Entry).TypedRune(0xc4201b6e10, 0xc400000073)
	/Users/pbarnum/go/src/fyne.io/fyne/widget/entry.go:232 +0x8d
fyne.io/fyne/driver/gl.(*window).charModInput(0xc420148200, 0xc420148180, 0x73, 0x0)
	/Users/pbarnum/go/src/fyne.io/fyne/driver/gl/window.go:807 +0x7c
fyne.io/fyne/driver/gl.(*window).(fyne.io/fyne/driver/gl.charModInput)-fm(0xc420148180, 0x73, 0x0)
	/Users/pbarnum/go/src/fyne.io/fyne/driver/gl/window.go:867 +0x46
fyne.io/fyne/vendor/github.com/go-gl/glfw/v3.2/glfw.goCharModsCB(0x6d00770, 0x73)
	/Users/pbarnum/go/src/fyne.io/fyne/vendor/github.com/go-gl/glfw/v3.2/glfw/input.go:300 +0x5d
fyne.io/fyne/vendor/github.com/go-gl/glfw/v3.2/glfw._cgoexpwrap_471f26bf8a94_goCharModsCB(0x6d00770, 0x73)
	_cgo_gotypes.go:2101 +0x3b
fyne.io/fyne/vendor/github.com/go-gl/glfw/v3.2/glfw._Cfunc_glfwPollEvents()
	_cgo_gotypes.go:1111 +0x41
fyne.io/fyne/vendor/github.com/go-gl/glfw/v3.2/glfw.PollEvents()
	/Users/pbarnum/go/src/fyne.io/fyne/vendor/github.com/go-gl/glfw/v3.2/glfw/window.go:777 +0x20
fyne.io/fyne/driver/gl.(*gLDriver).runGL(0xc4201211a0)
	/Users/pbarnum/go/src/fyne.io/fyne/driver/gl/loop.go:77 +0x4d2
fyne.io/fyne/driver/gl.(*gLDriver).Run(0xc4201211a0)
	/Users/pbarnum/go/src/fyne.io/fyne/driver/gl/driver.go:58 +0x43
fyne.io/fyne/driver/gl.(*window).ShowAndRun(0xc420148200)
	/Users/pbarnum/go/src/fyne.io/fyne/driver/gl/window.go:303 +0x57
main.main()
```

To reproduce:
1. Create a text input and button. On button click, run `input.SetText("")`
2. When app is running, type some characters into the text input
3. Click the button, this should set the input value to an empty string
4. Click in the text input to bring it into focus, you should see the cursor is not at the starting position, but slightly to the right of it.
5. Type a character
6. The app panics with the above call stack

Note: You can prevent the app from crashing, before step 5, hit the right/left arrow key and it should correct itself.

--- a/widget/entry.go
+++ b/widget/entry.go
@@ -167,8 +167,10 @@ func (e *Entry) SetText(text string) {
 	e.updateText(text)
 
 	if text == "" {
+		e.Lock()
 		e.CursorColumn = 0
 		e.CursorRow = 0
+		e.Unlock()
 		Renderer(e).(*entryRenderer).moveCursor()
 	}
 }

--- a/widget/entry.go
+++ b/widget/entry.go
@@ -165,6 +165,12 @@ func (e *Entry) Hide() {
 func (e *Entry) SetText(text string) {
 	e.textProvider().SetText(text)
 	e.updateText(text)
+
+	if text == "" {
+		e.CursorColumn = 0
+		e.CursorRow = 0
+		Renderer(e).(*entryRenderer).moveCursor()
+	}
 }
 
 // SetPlaceHolder sets the text that will be displayed if the entry is otherwise empty

--- a/widget/entry_test.go
+++ b/widget/entry_test.go
@@ -66,6 +66,29 @@ func TestEntry_SetPlaceHolder(t *testing.T) {
 	assert.Equal(t, 2, entry.textProvider().len())
 }
 
+func TestEntry_SetTextEmptyString(t *testing.T) {
+	entry := NewEntry()
+
+	assert.Equal(t, 0, entry.CursorColumn)
+
+	test.Type(entry, "test")
+	assert.Equal(t, 4, entry.CursorColumn)
+	entry.SetText("")
+	assert.Equal(t, 0, entry.CursorColumn)
+
+	entry = NewMultiLineEntry()
+	test.Type(entry, "test\ntest")
+
+	down := &fyne.KeyEvent{Name: fyne.KeyDown}
+	entry.TypedKey(down)
+
+	assert.Equal(t, 4, entry.CursorColumn)
+	assert.Equal(t, 1, entry.CursorRow)
+	entry.SetText("")
+	assert.Equal(t, 0, entry.CursorColumn)
+	assert.Equal(t, 0, entry.CursorRow)
+}
+
 func TestEntry_OnKeyDown(t *testing.T) {
 	entry := NewEntry()
 

